Centralized Transaction Management Infrastructure

anarazel opened the pull request:
See: #704

This is based ontop of #863. Thus until that's merged, it'll show superflous changes.

Since the version posted in #775 I've changed the following:
- introduced an explicit list of "in-progress" transactions as desired by @marcocitus 
- Moved 'critical' markers from connection to transaction, seems more accurate.
- 2PC handling is now triggered by CoordinatedTransactionUse2PC instead of MultiShardCommitProtocol. That'll allow to use to the transaction framework easily for reference table 2PC transactions, which don't properly fall under MultiShardCommitProtocol.
- Added follow partial conversion of multi_shard_transaction.[ch] to new framework. It now doesn't need its own callback anymore.
- Added followup partial conversion of router executor to new framework. Router now doesn't have it's own transaction callback anymore, but gets called back by transaction_management.c after the remote commit, to invalidate shards and such

In the future I'd hope to reuse transaction_recovery.[ch] to provide recovery for 2PC transactions. But that seems out of scope of this individual change.

Testing around this also revealed that we desperately need variants of PQexec et al (i.e. all blocking libpq functions) that do *not* block inside libpq, but instead do so in citus code. That will allow us to do always accept interrupts. That's especially important because it's still very easy to deadlock citus - which currently will sleep ininterruptible in many cases.

I don't think this is 100% ready, but I think it's more than close enough for some decent feedback.

marcocitus commented on src/backend/distributed/transaction/remote_transaction.c:
```diff
@@ -0,0 +1,611 @@
+
+			transaction->transactionState = REMOTE_TRANS_2PC_COMMITTING;
+
+			if (!SendRemoteCommand(connection, command.data))
```
It might be nice to encapsulate the commit/prepare/abort logic for individual connections into separate functions. There are several cases in which we will want to commit a connection prior to the coordinated transaction committing and it would be nice if we could use those functions.

anarazel replied:
I didn't do that initially because we should really wait on all the connections in parallel. But I think we can do that by having
```
extern void StartRemoteTransactionCommit(struct MultiConnection *connection);
extern void FinishRemoteTransactionCommit(struct MultiConnection *connection);
extern void RemoteTransactionCommit(struct MultiConnection *connection);
```
where ```StartRemoteTransactionCommit``` and ```FinishRemoteTransactionCommit``` are used by the central handler and ```RemoteTransactionCommit``` just wraps those.  To make that more efficient we later can add a "WaitForIOOnAll(List *connections)" or such.

Will do that.

marcocitus commented on src/backend/distributed/transaction/remote_transaction.c:
```diff
@@ -0,0 +1,611 @@
+			 * first.
+			 */
+			ForgetResults(connection); /* try to clear pending stuff */
+			if (!SendRemoteCommand(connection, "ROLLBACK;"))
```
Do we need the semicolon? Doesn't that end up executing two queries (one of which is empty).

anarazel replied:
Nope, removed all.

anarazel commented on src/include/distributed/connection_management.h:
```diff
@@ -61,7 +62,14 @@ typedef struct MultiConnection
+	RemoteTransaction remoteTransaction;
+
+	/* membership in list of in-progress transactions */
+	dlist_node transactionNode;
```
Yep.

anarazel replied:
(moving to RemoteTransaction makes access to the connection harder, because you then need to compute the offset to the MultiConnection and do pointer math (i.e. containerof()))

anarazel commented on src/backend/distributed/transaction/remote_transaction.c:
```diff
@@ -0,0 +1,611 @@
+ * parallel, to lessen latency penalties.
+ */
+void
+AdjustRemoteTransactionStates(List *connectionList)
```
I can work with the name, not happy about the !InCoordinatedTransaction, that'll be strewn all over.

marcocitus replied:
Still would like this to be something like `InitializeRemoteTransaction`. Adjust.. feels like we're just fixing up a few variables (like `AdjustColumnOldAttributes`, `AdjustStateForFailure`).

marcocitus replied:
Either one works for me, or `BeginRemoteTransactionBlock(s)IfNecessary`.

anarazel commented on src/backend/distributed/transaction/remote_transaction.c:
```diff
@@ -0,0 +1,611 @@
+		PGresult *result = NULL;
+
+		if (transaction->transactionFailed ||
+			transaction->transactionState != REMOTE_TRANS_STARTING)
```
Good plan.

anarazel commented on src/backend/distributed/transaction/remote_transaction.c:
```diff
@@ -0,0 +1,611 @@
+		result = PQgetResult(connection->pgConn);
+		if (!IsResponseOK(result))
+		{
+			ReportResultError(connection, result, WARNING);
```
Hm, ok. So we need a function that can "enrich" remote error messages.

anarazel commented:
This is a plain rebase over the merged connection-lifecycle PR, so this is easier to view. No other changes.

anarazel commented on src/include/distributed/remote_transaction.h:
```diff
@@ -0,0 +1,81 @@
+										bool allowErrorPromotion);
+extern void MarkRemoteTransactionCritical(struct MultiConnection *connection);
+
+extern void CoordinatedRemoteTransactionsCommit(void);
```
The namespace is currently a bit overloaded because there's CommitRemoteTransactions(). That's why I added the Coordinated() bit.

anarazel replied:
I've not renamed that yet due to that.   There's some argument to be made that these routines now could be moved into transaction_management.c, now that individual commits/rollbacks/... can be issued from other TUs. But that'd mean moving a fair amount of knowledge about the state of individual transactions, so I've refrained from that so far.

anarazel commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -75,11 +148,25 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
+			/*
+			 * FIXME: Add warning for the COORD_TRANS_COMMITTED case. That
+			 * can be reached if this backend fails after the
+			 * XACT_EVENT_PRE_COMMIT state.
```
It's entirely possible if multiple commit handlers exist...

marcocitus replied:
Fair enough.

anarazel commented:
@marcocitus I pushed a commit addressing *part* of your feedback. It's probably not worth looking over again in a lot of detail yet, but a quick comment about the split commit/abort/prepare bit would be appreciated (BEGIN should be done similarly).

With his split we should eventually be able to even make something like SendCommandListToWorkerInSingleTransaction() use the new framework.  I think it's pretty powerful to be able to commit early, but have things automatically committed / rolled back and such if you don't.

Pretty much the only major missing feature is support for logging prepared transaction. It looks to me like it should be fairly easy to make transaction_recovery.[ch] reusable for this. Basically expose LogTransactionRecord.  There's the question of whether we want somehow want to use the groupid (not entirely clear how to build that mapping), or whether we want to expand host/port to that api alternatively.

Is there any argument for not making *ALL* 2PC commits logged? I don't think the performance impact is even remotely large enough for us to care.

I also wonder if it'd wouldn't be a good idea to do have a function that marks a prepared transaction as "committed". If we use heap_inplace_update to set that, we can even do it it in the XACT_EVENT_COMMIT handler.  That'll save a bunch of network back in forth when prepare transactions are frequently being usedd.

marcocitus commented on src/backend/distributed/transaction/remote_transaction.c:
```diff
@@ -0,0 +1,611 @@
+			continue;
+		}
+
+		result = PQgetResult(connection->pgConn);
```
Yep. Ignoring timeouts for the moment works for me.

marcocitus commented:
>  quick comment about the split commit/abort/prepare bit would be appreciated (BEGIN should be done similarly).

Yep, I like this approach a lot.

> Is there any argument for not making ALL 2PC commits logged? I don't think the performance impact is even remotely large enough for us to care.

Not really, I suppose that `pg_dist_transaction` might grow a fair bit if you do a lot of COPY's and never run `recover_prepared_transactions`. 

Unrelated implementation notes: We should lock `pg_dist_transaction` prior to sending prepare commands, since transaction recovery must not run while prepared transactions exist that were created by an ongoing transaction, seems this got lost in porting back from MX. We should also start including `GetLocalGroupId()` in the 2PC identifier to distinguish 2PCs started by workers.

> I also wonder if it'd wouldn't be a good idea to do have a function that marks a prepared transaction as "committed". If we use heap_inplace_update to set that, we can even do it it in the XACT_EVENT_COMMIT handler. That'll save a bunch of network back in forth when prepare transactions are frequently being usedd.

`recover_prepared_transactions` fetches the list of pending prepared transactions from the worker and then compares it to the local list. Having a large number of committed transactions doesn't necessarily increase network traffic, but if doing a `heap_inplace_update` allows us to clean up `pg_dist_transaction` without running `recover_prepared_transactions` then that would be nice.

anarazel commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -75,11 +148,25 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
+			}
+
 			/* close connections etc. */
 			if (CurrentCoordinatedTransactionState != COORD_TRANS_NONE)
```
I'd rather make sure that we reset per-transaction state in any case.  Not that important though, if you feel strongly.

marcocitus replied:
Nope, not strongly.

anarazel commented on src/backend/distributed/connection/connection_management.c:
```diff
@@ -630,6 +633,11 @@ AfterXactHostConnectionHandling(ConnectionHashEntry *entry, bool isCommit)
+			/* reset per-transaction state */
+			connection->remoteTransaction.transactionState = REMOTE_TRANS_INVALID;
+			connection->remoteTransaction.transactionFailed = false;
+			connection->remoteTransaction.transactionCritical = false;
```
done.

anarazel commented:
> >  quick comment about the split commit/abort/prepare bit would be appreciated (BEGIN should be done similarly).
> 
> Yep, I like this approach a lot.

Great! I've done the same to Begin. Note that there's still Adjust* as
before, for when you don't know whether the connection already started a
transaction and/or whether a coordinated transaction is in progress.

> > Is there any argument for not making ALL 2PC commits logged? I don't think the performance impact is even remotely large enough for us to care.
> 
> Not really, I suppose that `pg_dist_transaction` might grow a fair bit if you do a lot of COPY's and never run `recover_prepared_transactions`. 

Yea. I was wondering if we shouldn't have a bgworker that deals with
doing so.  Difficulty is it would have to do so in any of the databases
:/

.oO(pg_cron)


> Unrelated implementation notes: We should lock `pg_dist_transaction`
> prior to sending prepare commands, since transaction recovery must not
> run while prepared transactions exist that were created by an ongoing
> transaction, seems this got lost in porting back from MX. We should
> also start including `GetLocalGroupId()` in the 2PC identifier to
> distinguish 2PCs started by workers.

I think it might be worth creating an issue for that.

I wonder if we couldn't do something based on xids about that. Allow
recovery to run concurrently, but only on rows that are guaranteed to be
visible for everyone - that's easy enough to do from C. Roughly like
```
if (TransactionIdPrecedes(HeapTupleHeaderGetXmin(heapTuple), GetOldestXmin(pg_dist_trans, true)))
{
    /* can perform cleanup, hasn't been written to concurrently */
}
```


> `recover_prepared_transactions` fetches the list of pending prepared
> transactions from the worker and then compares it to the local
> list. Having a large number of committed transactions doesn't
> necessarily increase network traffic, but if doing a
> `heap_inplace_update` allows us to clean up `pg_dist_transaction`
> without running `recover_prepared_transactions` then that would be
> nice.

We'd need a fixed-width 'status' field or such. Only those can be
updated inplace. But that seems like a small hurdle.


I've updated the PR, tried to address most of your comments, and so
other cleanup.  What do you think?

anarazel commented:
It's hard to increase coverage here without isolationtester tests. @jasonmp85 any comments about what it'd take to allow that (e.g. to run tests against a self-compiled postgres)? That'd also allow #1017 to run automatedly.

anarazel commented:
@marcocitus (and @sumedhpathak) We could convert worker_transaction.c to the new framework, which would bring coverage up again. But that's a relatively verbose change because we'd have to change the type of TransactionConnection->connection, which is still fairly widely used.  So I think I'd rather do that as a separate PR?

marcocitus commented on src/backend/distributed/connection/remote_commands.c:
```diff
@@ -184,12 +186,187 @@ LogRemoteCommand(MultiConnection *connection, const char *command)
+	/* make sure command has been sent out */
+	while (!failed)
+	{
+		int rc;
```
should be `int rc = 0;` I guess

marcocitus commented:
Not too worried about reducing code coverage with this PR.

marcocitus commented:
> We can't really throw here, because we want to commit other transactions etc. So I'm not really sure how we can do much better :(. I mean we could collect all failures, and then re-throw them in one hunk at the end, but that'd not be nice either. Maybe I'm missing something?

Not for commit or rollback, but we can directly throw an error for prepare transaction or begin.

anarazel commented:
On December 9, 2016 2:59:46 AM PST, Marco Slot <notifications@github.com> wrote:
>> We can't really throw here, because we want to commit other
>transactions etc. So I'm not really sure how we can do much better :(.
>I mean we could collect all failures, and then re-throw them in one
>hunk at the end, but that'd not be nice either. Maybe I'm missing
>something?
>
>Not for commit or rollback, but we can directly throw an error for
>prepare transaction or begin.

I don't think so - consider dropping a distributed table or such. We'll throw an error if it's a critical xact, but otherwise we gotta march on.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

anarazel commented on src/backend/distributed/connection/remote_commands.c:
```diff
@@ -184,12 +186,187 @@ LogRemoteCommand(MultiConnection *connection, const char *command)
+
+			/*
+			 * If raising errors allowed, or called within in a section with
+			 * interrupts held, return NULL instead.
```
Hm, we probably should additionally mark the transaction as failed - that should work well I think?

marcocitus replied:
Yep, makes sense. We should be careful that there are a few cases where rollback on cancellation is currently undesirable (master_append_table_to_shard), but that should fall outside the scope of the coordinated transaction.

marcocitus replied:
"currently undesirable" as in:  the current implementation sucks :). 2PC would be much better there. I guess there's a trade-off between master_append_table_to_shard getting stuck and having inconsistent shard placements, which could ultimately lead to data loss. I think this function getting stuck might be the lesser of 2 evils, but probably depends on who you ask.

anarazel replied:
Ah ;).  Yea, it won't affect master_append_table_to_shard as-is.  But I don't think master_append_table_to_shard defends against being stuck in a meaningful manner right now anyway? It holds interrupts while setting placement status, but not while actually doing the amending, which already committed at that point anyway?   Brrr.

anarazel commented:
@marcocitus Think the ball's in your court now ;)

marcocitus commented on src/backend/distributed/connection/remote_commands.c:
```diff
@@ -52,7 +54,7 @@ ForgetResults(MultiConnection *connection)
 	{
 		PGresult *result = NULL;
-		result = PQgetResult(connection->pgConn);
+		result = GetRemoteCommandResult(connection, false);
```
It's worth following our convention of using a raiseInterrupts variable as an argument such that the reader knows what the parameter does without looking at the function.

jasonmp85 replied:
> Done. But god, I hate that policy.

Like many such policies, I feel this is compensating for poor tools. If I see a `false` there, a hover affordance can show me the name of the variable in the function prototype. I'm certain any other well-configured editor has something similar.

marcocitus commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -1902,8 +1764,13 @@ MarkRemainingInactivePlacements(void)
+			 * failed).
+			 */
+			if (participant->connection == NULL ||
+				participant->connection->remoteTransaction.transactionFailed)
```
Maybe extract the connection variable.

anarazel replied:
Done.

marcocitus commented on src/backend/distributed/transaction/multi_shard_transaction.c:
```diff
@@ -167,12 +170,14 @@ BeginTransactionOnShardPlacements(uint64 shardId, char *userName)
+		 * Every individual failure should cause entire distributed
+		 * transaction to fail.
+		 */
+		MarkRemoteTransactionCritical(connection);
```
Shouldn't this be set before the begin? 

Also, what if we call this automatically using something like a `TRANSACTION_CRITICAL` flag?

anarazel replied:
> Shouldn't this be set before the begin?
It should be set before the begin, wouldn't make much of a difference though, we'll fail just a tiny bit later. Will change.

> Also, what if we call this automatically using something like a TRANSACTION_CRITICAL flag?

I wondered about that, but decided that it's not that unlikely that a) the need to set the flag only arises after having interactions with the remote b) it's not known to the connection establishing layer that it needs to be set.

So I'm slightly inclined to leave this as-is.

marcocitus commented:
Added some optional stylistic comments and I think there's a minor bug in BeginTransactionOnShardPlacements, but looks ready to 🚢  after fixing it.

anarazel commented:
On 2016-12-12 15:19:19 -0800, Jason Petersen wrote:
> jasonmp85 commented on this pull request.
> 
> 
> 
> > @@ -52,7 +54,7 @@ ForgetResults(MultiConnection *connection)
>  	while (true)
>  	{
>  		PGresult *result = NULL;
> -		result = PQgetResult(connection->pgConn);
> +		result = GetRemoteCommandResult(connection, false);
> 
> > Done. But god, I hate that policy.
> 
> Like many such policies, I feel this is compensating for poor tools. If I see a `false` there, a hover affordance can show me the name of the variable in the function prototype. I'm certain any other well-configured editor has something similar.

Indeed. *emacs* a tool from before the dark ages can do so.

anarazel commented:
@marcocitus Thanks for the review! Merging. If we want to reconsider on the critical thing, we can easily do so in a smaller PR.
